keps/sig-cli/20190318-kustomize-generators-transformers.md

```diff
@@ -0,0 +1,389 @@
+  - Having the groupVersion `helm.kustomize.io` triggers the generator under `plugins/generators/helm.kustomize.io`
+
+```yaml
+groupVersion: helm.kustomize.io/v1beta1
+kind: Chart
+generate:
+    chart: ./path/to/chart/from/kustomization.yaml
+    values:
+      k1: v1
+      k2: v2
+```
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+generators:
+- chart.yaml
+```
+
+The `helm.kustomize.io` binary is invoked with the `chart.yaml` contents passed on STDIN.  Kustomize adds the emitted Resource
+to its set of Resources.
+
+#### Bespoke Abstractions for Organizations
+
+Bob's organization deploys many variants of the same high-level conceptual set of Resources.  Creating the Resource Config
+for each instance that needs to be deployed requires lots of boilerplate.  Instead Bob's organization develops
+tools for generating standardized Resource Config based off of some inputs.
+
+Bob creates a new generator for his organization that allow higher level abstractions to be defined as
+new virtual Resource types that don't exist in the cluster, but are used to generate low-level types.
+
+1. Bob builds and installs the new generator plugin `$XDG_CONFIG_HOME/kustomize/plugins/generators/team.example.com`.
+1. Bob creates the `app.yaml` Resource Config and adds it to his `kustomization.yaml`
+
+```yaml
+groupVersion: team.example.com/v1beta1
+kind: CommonApp
+generate:
+  image: foo
+  size: big
+```
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+generators:
+- app.yaml
+```
+
+The `team.example.com` binary is invoked with the `app.yaml` contents passed on STDIN.  Kustomize adds the
+emitted Resource to its set of Resources.
+
+If multiple generators are specified, they are run in the order they are listed.
+
+#### Bespoke Configuration for Organizations
+
+Alice's Organization requires that various fields are defaulted if unset.  SRE would like to be able to see
+the full Resources that are being Applied and have this auditable in an scm, such as git, rather than
+having Webhooks provide server-side mutions that are not capture in review or scm.
+
+1. Alice builds and installs the new transformer plugin at `$XDG_CONFIG_HOME/kustomize/plugins/transformers/team.example.com`
+1. Alice creates the `transform.yaml` Resource Config and adds it to her `kustomization.yaml` as a `transformers` entry.
+
+```yaml
+groupVersion: team.example.com
+kind: CommonApp
+transform:
+  namePrefix: foo-
+  commonAnnotations:
+    foo: bar
+```
+
+```yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+transformers:
+- transform.yaml
+```
+
+The `team.example.com` binary is invoked with both the `transform.yaml` contents and the Resources passed on STDIN.
+Kustomize replaces its Resources with the emitted Resources.
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+## Design Details
```

Comment:
Are plugins expected to produce stable/repeatable output?

Do the plugins get executed during `--dry-run`?  If not, is there another mode that does execute them but doesn't apply the results?

Reply:
Also, just for my own clarity: if a kustomization.yaml references several remote bases that also use kustomization - I presume each remote base is "expanded" using the generators/transformers specified in that remote kustomization.yaml (but using local plugin executables).  And then the result of each base is concatenated together before passing to the first locally-specified transformer plugin.

Iow, we're building a tree of plugin pipelines here, not just a single linear pipeline.

Reply:
> Do the plugins get executed during --dry-run

Yes, `kustomize` doesn't have a notion of dry-run.  Could be proposed in a separate KEP.

> ...Iow, we're building a tree of plugin pipelines here, not just a single linear pipeline.

Yes, this is how Kustomize works today.

Reply:
> Yes, kustomize doesn't have a notion of dry-run

kubectl does though.  So if we're proposing that these features become part of the kustomization.yaml support within kubectl, then we need to decide as part of implementing this KEP ...

Reply:
Good point.

Essentially you are asking to clarify that bases can define `generator` and `transformer` *Virtual Resources*, and these would be lazily invoked when kustomize was run.  Is this correct?

Reply:
I think we now have two useful versions of --dry-run, and I was just looking for a statement which one we're planning to do here:

- Run no plugins at all:
  No plugins are executed, but `kubectl create --dry-run -o yaml` does not show the objects/transformations that will actually exist without `--dry-run`
- Run plugins but don't push to server:
  Plugins still get executed, and are just assumed to not have permanent undesirable side-effects (which is why we should document it here).
  `kubectl create --dry-run -o yaml` shows something similar to the objects that would be created without `--dry-run` (assuming plugins are mostly repeatable).

There are issues with both, so it would be good to discuss/agree on it at this stage rather than be surprised by the implementation.

Reply:
I changed this to say plugins must behave as if they are always run in dry-run

Reply:
> Plugins should probably accept the dry-run flag

The current KEP doesn't have any notion of "common arguments" provided by kustomize itself (rather than from the user-provided virtual resource).  I'm guessing in the future we will want something like this.

I think we will have to use environment variables for this (which is fine) unless we want to change this KEP now to include an "envelope metadata" wrapper around the virtual resource provided on stdin.  We also don't have a versioned-api at all (other than changing the install location) to allow us to change the plugin communication in incompatible ways in future.

I think we can work around that, so mostly just raising it as a future-cost.

Reply:
SGTM.  We will need to figure out which ones will be required for the plugin to support (these probably need to be set as flags so the command fails), vs are optional (these probably need to be env vars so the command doesn't value).  Plugins will probably want to use something like Viper to read configuration from flags and env vars arbitrarily.